src/ring_theory/perfection.lean

```diff
@@ -113,7 +111,115 @@ lemma coeff_ne_zero_of_le {f : ring.perfection R p} {m n : ℕ} (hfm : coeff R p
   (hmn : m ≤ n) : coeff R p n f ≠ 0 :=
 let ⟨k, hk⟩ := nat.exists_eq_add_of_le hmn in hk.symm ▸ coeff_add_ne_zero hfm k
 
-end ring.perfection
+variables (R p)
+
+instance perfect_ring : perfect_ring (ring.perfection R p) p :=
+{ pth_root' := pth_root R p,
+  frobenius_pth_root' := congr_fun $ congr_arg ring_hom.to_fun $ @frobenius_pth_root R _ p _ _,
+  pth_root_frobenius' := congr_fun $ congr_arg ring_hom.to_fun $ @pth_root_frobenius R _ p _ _ }
+
+instance ring (R : Type u₁) [comm_ring R] [char_p R p] : ring (ring.perfection R p) :=
+((ring.perfection R p).to_subring $ λ n, by simp_rw [← frobenius_def, pi.neg_apply,
+    pi.one_apply, ring_hom.map_neg, ring_hom.map_one]).to_ring
+
+instance comm_ring (R : Type u₁) [comm_ring R] [char_p R p] : comm_ring (ring.perfection R p) :=
+((ring.perfection R p).to_subring $ λ n, by simp_rw [← frobenius_def, pi.neg_apply,
+    pi.one_apply, ring_hom.map_neg, ring_hom.map_one]).to_comm_ring
+
+/-- Given rings `R` and `S` of characteristic `p`, with `R` being perfect,
+any homomorphism `R →+* S` can be extended to a homomorphism `R →+* perfection S p`. -/
+@[simps] def extend (R : Type u₁) [comm_semiring R] [char_p R p] [perfect_ring R p]
+  (S : Type u₂) [comm_semiring S] [char_p S p] :
+  (R →+* S) ≃ (R →+* ring.perfection S p) :=
+{ to_fun := λ f,
+  { to_fun := λ r, ⟨λ n, f $ _root_.pth_root R p ^[n] r,
+      λ n, by rw [← f.map_pow, function.iterate_succ_apply', pth_root_pow_p]⟩,
+    map_one' := ext $ λ n, (congr_arg f $ ring_hom.iterate_map_one _ _).trans f.map_one,
+    map_mul' := λ x y, ext $ λ n, (congr_arg f $ ring_hom.iterate_map_mul _ _ _ _).trans $
+      f.map_mul _ _,
+    map_zero' := ext $ λ n, (congr_arg f $ ring_hom.iterate_map_zero _ _).trans f.map_zero,
+    map_add' := λ x y, ext $ λ n, (congr_arg f $ ring_hom.iterate_map_add _ _ _ _).trans $
+      f.map_add _ _ },
+  inv_fun := ring_hom.comp $ coeff S p 0,
+  left_inv := λ f, ring_hom.ext $ λ r, rfl,
+  right_inv := λ f, ring_hom.ext $ λ r, ext $ λ n,
+    show coeff S p 0 (f (_root_.pth_root R p ^[n] r)) = coeff S p n (f r),
+    by rw [← coeff_iterate_frobenius _ 0 n, zero_add, ← ring_hom.map_iterate_frobenius,
+      right_inverse_pth_root_frobenius.iterate] }
+
+end perfection
+
+/-- A perfection map to a ring of characteristic `p` is a map that is isomorphic
+to its perfection. -/
+@[nolint has_inhabited_instance] structure perfection_map (p : ℕ) [fact p.prime]
```

Comment:
Should this be a structure extending ring homs, or a predicate on top of ring homs?
For `localization_map` we chose to extend, and I think we are now regretting it. (Indeed Riccardo has been looking into refactoring it.)
For example, if we choose the predicate version, then it's a lot easier to mix with `algebra P R`.

Reply:
I'm just afraid that in the Year 2030 our assumptions will be a 37 lines long...

Reply:
I'm sure we can find technical solutions to that in the next 5 years.

Reply:
Done.

Reply:
Thanks! Should it be in the `ring_hom` namespace? I don't know if we want to write `f.perfection_map` for the predicate?

Reply:
The drawback is that if we don't want to specify the hom then we would have to write `ring_hom.perfection_map _` as opposed to the alternative `perfection_map _`. I think I prefer having it in the global namespace.

Reply:
okay, fair enough